asyncapi/identity.server/dispatcher_service.bal

```diff
@@ -0,0 +1,217 @@
+    isolated function removeServiceRef(string serviceType) returns error? {
+        if (!self.services.hasKey(serviceType)) {
+            return error("Cannot detach the service of type " + serviceType +
+            ". Service has not been attached to the listener before");
+        }
+        _ = self.services.remove(serviceType);
+    }
+
+    // We are not using the (@http:payload GenericEventWrapperEvent g) notation because of a bug in Ballerina.
+    // Issue: https://github.com/ballerina-platform/ballerina-lang/issues/32859
+    resource function post .(http:Caller caller, http:Request request) returns http:Response|error? {
+        json payload = check request.getJsonPayload();
+        byte[] binaryPay = payload.toString().toBytes();
+        string signature = check request.getHeader("x-wso2-event-signature");
+        string trimmedSecret = signature.substring(7, signature.length());
+        byte[] output = check crypto:hmacSha256(binaryPay, self.listenerConfig.webhookSecret.toBytes());
+        string computedDigest = output.toBase16();
+        if (trimmedSecret.length() != computedDigest.length() || trimmedSecret !== computedDigest) {
+            // Validate secret with x-wso2-event-signature header for intent verification.
+            log:printError("Signature verification failure");
+            http:Response response = new;
+            response.statusCode = http:STATUS_UNAUTHORIZED;
+            response.setPayload("Signature verification failure");
+            return response;
+        }
+
+        GenericPayloadType|error genericPayloadType = payload.cloneWithType();
+        if (genericPayloadType is error) {
+            log:printError("Unsupported payload");
+            return genericPayloadType;
+        } else {
+            check self.matchRemoteFunc(genericPayloadType);
+            http:Response response = new;
+            response.statusCode = http:STATUS_OK;
+            response.setPayload("Event acknowledged successfully");
+            return response;
+        }
+    }
+
+    private function matchRemoteFunc(GenericPayloadType genericPayloadType) returns error? {
+        json events = genericPayloadType.events;
+        map<json> eventMap = <map<json>>events;
+        string eventKey = eventMap.keys()[0];
+        regexp:RegExp eventPattern = check regexp:fromString("/events/");
+        string mainEventWithEventType = regexp:split(eventPattern, eventKey)[1];
+        regexp:RegExp eventTypePattern = check regexp:fromString("/event-type/");
+        string[] eventDetails = regexp:split(eventTypePattern, mainEventWithEventType);
+        string mainEvent = eventDetails[0];
+        string eventType = eventDetails[1];
+
+        match mainEvent {
+            "user" => {
+                match eventType {
+                    "userCreated" => {
+                        json eventData = eventMap[eventKey];
+                        GenericUserOperationData userCreatedData = check eventData.cloneWithType();
```

Comment:
There seems to be code blocks repeated across clauses? Shall we refactor the code to avoid that?

Reply:
Fixed with https://github.com/ballerina-platform/asyncapi-triggers/pull/171/commits/da22e6a64070c1f3ccb984dee547ed5039100085